types/types.go

```diff
@@ -1239,6 +1231,175 @@ func (res *V2FileContractResolution) UnmarshalJSON(b []byte) error {
 	if err := json.Unmarshal(p.Resolution, res.Resolution); err != nil {
 		return err
 	}
-	res.Parent = p.Parent
+	res.Parent = p.Parent.Move()
 	return nil
 }
+
+// MarshalJSON implements json.Marshaller.
+//
+// For convenience, the transaction's ID is also calculated and included. This
+// field is ignored during unmarshalling.
+func (txn V2Transaction) MarshalJSON() ([]byte, error) {
+	type jsonTxn V2Transaction // prevent recursion
+	return json.Marshal(struct {
+		ID TransactionID `json:"id"`
+		jsonTxn
+	}{txn.ID(), jsonTxn(txn)})
+}
+
+// To guard against memory ownership bugs, all Element types have Move, Share,
+// and Copy methods. This enables a linter to flag any instances where Element
+// memory is not explicitly managed.
+
+// CanMutate returns whether the element's memory can safely be mutated.
+func (se StateElement) CanMutate() bool { return !se.shared }
```

Comment:
All of this looks promising but this makes me a bit sad. It means you still need to think about calling `CanMutate` before mutating which is something you might forget.

The only way around this is probably to have a container type around StateElement.

```
type StateElementContainer struct {
  stateElement *StateElement
  shared bool
}

func (c *StateElementContainer) MutRef() *StateElement {
  if c.shared {
    panic("bad stuff")
  }
  return c.stateElement
}
```

In this case you wouldn't even be able to get a `StateElement` to mutate without calling `MutRef` to get a mutable reference. Still not foolproof and probably 10 times more nasty.

Reply:
`CanMutate` only really exists to be used by `UpdateElementProof`. I suppose it might not be needed, though; the `Move` method is kinda like your `MutRef`, in that it will panic if called on a shared element. So we can do:
```go
func UpdateElementProof(e *types.StateElement) {
    _ = e.Move() // panics if e is shared
    // ...
}
```
The panic message won't be as nice, but it's still an improvement over having a separate method just for one caller.